File: app/src/main/java/com/bitchat/android/model/RequestSyncPacket.java
package com.bitchat.android.model;

import com.bitchat.android.sync.SyncDefaults;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Arrays;
import java.util.Objects;

/**
 * Paquet de synchronisation (REQUEST_SYNC) utilisant les paramètres GCS (Golomb-Coded Set).
 * Encodage TLV (type, longueur sur 16 bits, valeur) pour les champs.
 */
public final class RequestSyncPacket {

    public static final int MAX_ACCEPT_FILTER_BYTES = SyncDefaults.MAX_ACCEPT_FILTER_BYTES;

    private final int p;
    private final long m;
    private final byte[] data;

    public RequestSyncPacket(int p, long m, byte[] data) {
        this.p = p;
        this.m = m;
        this.data = data;
    }

    // Getters
    public int getP() { return p; }
    public long getM() { return m; }
    public byte[] getData() { return data; }

    /**
     * Encode le paquet en données binaires TLV.
     * @return Le tableau de bytes encodé.
     */
    public byte[] encode() {
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        try {
            // P
            putTLV(buffer, 0x01, new byte[]{(byte) p});
            // M (uint32)
            long m32 = Math.min(m, 0xffff_ffffL);
            byte[] mBytes = new byte[]{
                (byte) ((m32 >> 24) & 0xFF),
                (byte) ((m32 >> 16) & 0xFF),
                (byte) ((m32 >> 8) & 0xFF),
                (byte) (m32 & 0xFF)
            };
            putTLV(buffer, 0x02, mBytes);
            // data
            putTLV(buffer, 0x03, data);
        } catch (IOException e) {
            // Ne devrait jamais arriver avec ByteArrayOutputStream
            return null;
        }
        return buffer.toByteArray();
    }

    private void putTLV(ByteArrayOutputStream buffer, int type, byte[] value) throws IOException {
        buffer.write((byte) type);
        int len = value.length;
        buffer.write((byte) ((len >> 8) & 0xFF));
        buffer.write((byte) (len & 0xFF));
        buffer.write(value);
    }

    /**
     * Décode le paquet à partir de données binaires.
     * @param data Les données à décoder.
     * @return Un objet RequestSyncPacket, ou null en cas d'erreur.
     */
    public static RequestSyncPacket decode(byte[] data) {
        int off = 0;
        Integer p = null;
        Long m = null;
        byte[] payload = null;

        while (off + 3 <= data.length) {
            int t = data[off++] & 0xFF;
            int len = ((data[off++] & 0xFF) << 8) | (data[off++] & 0xFF);
            if (off + len > data.length) return null;
            byte[] v = Arrays.copyOfRange(data, off, off + len);
            off += len;

            switch (t) {
                case 0x01:
                    if (len == 1) p = v[0] & 0xFF;
                    break;
                case 0x02:
                    if (len == 4) {
                        m = ((long)(v[0] & 0xFF) << 24) |
                            ((long)(v[1] & 0xFF) << 16) |
                            ((long)(v[2] & 0xFF) << 8) |
                            ((long)(v[3] & 0xFF));
                    }
                    break;
                case 0x03:
                    if (v.length > MAX_ACCEPT_FILTER_BYTES) return null;
                    payload = v;
                    break;
            }
        }

        if (p == null || m == null || payload == null) return null;
        if (p < 1 || m <= 0L) return null;

        return new RequestSyncPacket(p, m, payload);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        RequestSyncPacket that = (RequestSyncPacket) o;
        return p == that.p && m == that.m && Arrays.equals(data, that.data);
    }

    @Override
    public int hashCode() {
        int result = Objects.hash(p, m);
        result = 31 * result + Arrays.hashCode(data);
        return result;
    }
}

File: app/src/main/java/com/bitchat/android/model/RequestSyncPacket.kt
package com.bitchat.android.model

import com.bitchat.android.protocol.TlvLengthSize
import com.bitchat.android.protocol.TlvReader
import com.bitchat.android.protocol.TlvWriter
import com.bitchat.android.protocol.UnknownTlvPolicy
import com.bitchat.android.sync.SyncDefaults

/**
 * REQUEST_SYNC payload using GCS (Golomb-Coded Set) parameters.
 * TLV (type, length16, value), types:
 *  - 0x01: P (uint8) — Golomb-Rice parameter
 *  - 0x02: M (uint32, big-endian) — hash range (N * 2^P)
 *  - 0x03: data (opaque) — GR bitstream bytes
 */
data class RequestSyncPacket(
    val p: Int,
    val m: Long,
    val data: ByteArray
) {
    fun encode(): ByteArray {
        val m32 = m.coerceAtMost(0xffff_ffffL)
        return TlvWriter()
            .put(0x01, byteArrayOf(p.toByte()), TlvLengthSize.TWO_BYTES)
            .put(
                0x02,
                byteArrayOf(
                    ((m32 ushr 24) and 0xFF).toByte(),
                    ((m32 ushr 16) and 0xFF).toByte(),
                    ((m32 ushr 8) and 0xFF).toByte(),
                    (m32 and 0xFF).toByte()
                ),
                TlvLengthSize.TWO_BYTES
            )
            .put(0x03, data, TlvLengthSize.TWO_BYTES)
            .toByteArray()
    }

    companion object {
        // Receiver-side safety limit (configurable constant)
        const val MAX_ACCEPT_FILTER_BYTES: Int = SyncDefaults.MAX_ACCEPT_FILTER_BYTES

        fun decode(data: ByteArray): RequestSyncPacket? {
            var p: Int? = null
            var m: Long? = null
            var payload: ByteArray? = null

            val fields = TlvReader.decode(
                data = data,
                defaultLengthSize = TlvLengthSize.TWO_BYTES,
                unknownPolicy = UnknownTlvPolicy.SKIP,
                knownTypes = setOf(0x01, 0x02, 0x03)
            ) ?: return null

            for (field in fields) {
                when (field.type) {
                    0x01 -> if (field.value.size == 1) p = (field.value[0].toInt() and 0xFF)
                    0x02 -> if (field.value.size == 4) {
                        val mm = ((field.value[0].toLong() and 0xFF) shl 24) or
                                 ((field.value[1].toLong() and 0xFF) shl 16) or
                                 ((field.value[2].toLong() and 0xFF) shl 8) or
                                 (field.value[3].toLong() and 0xFF)
                        m = mm
                    }
                    0x03 -> {
                        if (field.value.size > MAX_ACCEPT_FILTER_BYTES) return null
                        payload = field.value
                    }
                }
            }

            val pp = p ?: return null
            val mm = m ?: return null
            val dd = payload ?: return null
            if (pp < 1 || mm <= 0L) return null
            return RequestSyncPacket(pp, mm, dd)
        }
    }
}
